akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```
@@ -549,50 +549,61 @@ object HttpEntity {
 
-  /**
-   * INTERNAL API
-   */
-  private def limitable[Out, Mat](source: Source[Out, Mat], sizeOf: Out ⇒ Int): Source[Out, Mat] =
-    source.via(Flow[Out].transform { () ⇒
-      new PushStage[Out, Out] {
-        var maxBytes = -1L
-        var bytesLeft = Long.MaxValue
-
-        override def preStart(ctx: LifecycleContext) =
-          ctx.attributes.getFirst[SizeLimit] match {
-            case Some(limit: SizeLimit) if limit.isDisabled ⇒
-            // "no limit"
-            case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
-              if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
-            // else we still count but never throw an error
-            case Some(SizeLimit(bytes, None)) ⇒
-              maxBytes = bytes
-              bytesLeft = bytes
-            case None ⇒
-          }
-
-        def onPush(elem: Out, ctx: stage.Context[Out]): stage.SyncDirective = {
-          bytesLeft -= sizeOf(elem)
-          if (bytesLeft >= 0) ctx.push(elem)
-          else ctx.fail(EntityStreamSizeException(maxBytes))
+  private val limitableDefaults = Attributes.name("limitable")
+
+  private final class Limitable[T](sizeOf: T ⇒ Int) extends GraphStage[FlowShape[T, T]] {
+    val in = Inlet[T]("Limitable.in")
+    val out = Outlet[T]("Limitable.out")
+    override val shape = FlowShape.of(in, out)
+    override protected val initialAttributes: Attributes = limitableDefaults
+
+    override def createLogic(attributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var maxBytes = -1L
+      private var bytesLeft = Long.MaxValue
+
+      override def preStart(): Unit = {
+        attributes.getFirst[SizeLimit] match {
+          case Some(limit: SizeLimit) if limit.isDisabled ⇒
+          // "no limit"
+          case Some(SizeLimit(bytes, cl @ Some(contentLength))) ⇒
+            if (contentLength > bytes) throw EntityStreamSizeException(bytes, cl)
```

Looks weird to have two limits here. I guess we could fail fast for `contentLength > maxContentLength` in the blueprints but certainly out of scope here.